TPC space charge: first implementation of realistic space-charge propagation and modifications to store pre-calculated space-charge objects

ehellbar opened the pull request:
This pull request basically includes two modifications:

1) Implementation of a method to propagate space-charge inside the TPC and a slow method to propagate single ions as a validation cross check

2) Modifications in the GPU/TPCSpaceChargeBase classes' streamers, TPC digitizer and digitizer workflow to allow to store a pre-calculated space-charge object in a ROOT file and pass it to the workflow for digitization with distortions.

ehellbar commented:
...rebasing dev and fixing issues...
Cheers,
Ernst

shahor02 commented on GPU/TPCSpaceChargeBase/AliTPC3DCylindricalInterpolator.h:
```diff
@@ -46,16 +47,17 @@ class AliTPC3DCylindricalInterpolator
   Double_t* fPhiList = nullptr;    //[fNPhi] coordinate in phiList (rad) (should be increasing) 0 <= < 2 pi (cyclic)
   Double_t* fZList = nullptr;      //[fNZ] coordinate in z list (cm) (should be increasing)
-  Double_t* fSecondDerZ = nullptr; //[fNR*fNPhi*fNZ] store second derivative of cubic interpolation in z direction
+  Double_t* fSecondDerZ = nullptr; //[fNGridPoints] store second derivative of cubic interpolation in z direction
```
This is a non-virtual class, use ClassDefNV. More general point (just a suggestion): since you are anyway modifying these legacy classes, perhaps they can be brought to o2 conventions and new/delete policy? Why not use vectors here?

shahor02 commented on GPU/TPCSpaceChargeBase/AliTPC3DCylindricalInterpolatorIrregular.h:
```diff
@@ -130,8 +132,8 @@ class AliTPC3DCylindricalInterpolatorIrregular
-  KDTreeNode* fKDTreeIrregularPoints = nullptr; // to save tree as list
-  KDTreeNode* fKDTreeIrregularRoot = nullptr;   // kdtree root
+  KDTreeNode* fKDTreeIrregularPoints = nullptr; //!<![fNGridPoints] to save tree as list
+  KDTreeNode* fKDTreeIrregularRoot = nullptr;   //!<! kdtree root TODO: make this streamable
```
This class also should have ``ClassDefNV``

shahor02 commented on GPU/TPCSpaceChargeBase/AliTPCLookUpTable3DInterpolatorD.h:
```diff
@@ -65,17 +65,17 @@ class AliTPCLookUpTable3DInterpolatorD
-  Double_t* fZList = nullptr;   //[fNZ] List of Z coordinate (regular grid)
+  Double_t* fRList = nullptr;   //!<! List of R coordinate (regular grid)
+  Double_t* fPhiList = nullptr; //!<! List of Phi coordinate (regular grid)
+  Double_t* fZList = nullptr;   //!<! List of Z coordinate (regular grid)
```
Also ``ClassDefNV``

shahor02 commented on GPU/TPCSpaceChargeBase/AliTPCLookUpTable3DInterpolatorIrregularD.h:
```diff
@@ -80,17 +80,17 @@ class AliTPCLookUpTable3DInterpolatorIrregularD
-  TMatrixD** fMatricesZPoint = nullptr;   ///< Matrices to store distorted point (z component)
+  TMatrixD** fMatricesRPoint = nullptr;   //!<! Matrices to store distorted point (r component)
+  TMatrixD** fMatricesPhiPoint = nullptr; //!<! Matrices to store distorted point (phi component)
+  TMatrixD** fMatricesZPoint = nullptr;   //!<! Matrices to store distorted point (z component)
```
ClassDefNV

shahor02 commented on GPU/TPCSpaceChargeBase/AliTPCPoissonSolver.h:
```diff
@@ -122,7 +122,7 @@ class AliTPCPoissonSolver : public TNamed
   AliTPCPoissonSolver& operator=(const AliTPCPoissonSolver&); // not implemented
   StrategyType fStrategy = kMultiGrid;                        ///< strategy used default multiGrid
-  TMatrixD** fExactSolution = nullptr;                        ///< Pointer to exact solution
+  TMatrixD** fExactSolution = nullptr;                        //!<! Pointer to exact solution
```
ClassDefNV

shahor02 commented on GPU/TPCSpaceChargeBase/AliTPCSpaceCharge3DCalc.h:
```diff
@@ -440,8 +441,7 @@ class AliTPCSpaceCharge3DCalc
   /// \cond CLASSIMP
-  ClassDef(AliTPCSpaceCharge3DCalc,
-           1);
+  ClassDef(AliTPCSpaceCharge3DCalc, 1);
```
ClassDefNV

ehellbar commented:
Hi @shahor02 ,
I modified the functions you mentioned to be declared as const, used ClassDefNV wherever appropriate and changed the TMatrixD pointer to unique_ptrs in the TPCSimulation/SpaceCharge class. 

Concerning all the TMatrixD pointers and queries in the aliroot legacy classes and bringing them to O2 conventions, I prefer to do it in a separate PR. I agree that this should be done (light-weight 2d matrix class, std::vectors of matrices, etc.) but we have to synchronize it with Rifki. 

Cheers,
Ernst

shahor02 commented on Detectors/TPC/simulation/include/TPCSimulation/SpaceCharge.h:
```diff
@@ -61,98 +63,149 @@ class SpaceCharge
+  SpaceCharge(int nRBins, int nPhiBins, int nZSlices, int interpolationOrder);
 
   // Destructor
   ~SpaceCharge() = default;
```
@ehellbar OK for fixing legacy stuff in other PR, but this one was not addressed: you are using default destructor for a class which allocates lot of memory, this will lead to a memory leak. Could you please add a proper destructor?

ehellbar replied:
Hello @shahor02 ,
in this class (SpaceCharge.h), I put unique_ptrs where there were raw pointers before, so they all should be destroyed by the default constructor? 
Now I also checked in the legacy classes and modified one of their destructors accordingly.

shahor02 replied:
Hi @ehellbar 
thanks, yes, with unique_ptrs it should work, will merge once tests are passed.

wiechula commented:
@ehellbar : Thanks a lot for all the work
@shahor02 : Thanks for the comment
I think these classes are rather involved and nested in ways that are not that staight forward to understand. In mid-term it would be good to go through them again and try to simplify.
And as next step, as Ruben suggested, implement a lightweight matrix class. I thinks this also offers room for speedup in all the calculations.
